pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentselector/OfflineSegmentSelector.java

```diff
@@ -42,7 +49,14 @@ public void onExternalViewChange(ExternalView externalView, Set<String> onlineSe
     // TODO: for new added segments, before all replicas are up, consider not selecting them to avoid causing
     //       hotspot servers
 
-    _segments = Collections.unmodifiableList(new ArrayList<>(onlineSegments));
+
+    // Update segment lineage based segment selector
+    _segmentLineageBasedSegmentSelector.onExternalViewChange();
+
+    // Compute the intersection of segments to process from both offline and segment lineage based segment selectors.
+    List<String> segmentsToProcess =
+        new ArrayList<>(_segmentLineageBasedSegmentSelector.computeSegmentsToProcess(onlineSegments));
```

Comment:
I am unclear on the extensibility of the code in future. Will we add more selectors and cascade them in future, for example versionBasedSelector/Pruner?

Reply:
Another way is to model this as a `SegmentPreselector` which filters the `onlineSegments` before getting into the `SegmentSelector`

Reply:
thinking out loud here, should we model them it as pruner?

Reply:
@kishoreg Based on the interface definition, I feel it is better to model it as SegmentSelector (`The segments selected should cover the whole dataset (table) without overlap`). Another benefit of putting it in `Selector`/`Preselector` is to avoid the per-query calculation within the `Pruner`